### ISSUE TITLE
create_patches_reduced_target

### ISSUE THREAD

**Issue by lizard24** [CSBDeep/CSBDeep#43]
Hi,

first of all many thanks for creating such an easy-to-use framework for deep learning, I really appreciate it!

I have tried to train the network with 3D images as source/low, and 2D images as target/GT, but have trouble making it work. I thought this is what create_patches_reduced_target() is for but I might have misunderstood it.

I first save image pairs in folder "1" and "2", where images in folder "1" are in 3D (CYX) and in "2" in 2D (YX). I then create the rawdata object:

    raw_data = RawData.from_folder (
    basepath    = folder_out,
    source_dirs = ['1'],
    target_dir  = '2',
    axes        =  'CYX',
    )

and create patches:

        X, Y, XY_axes = create_patches_reduced_target(
            raw_data            = raw_data,
            patch_size          = (3,100,100),
            n_patches_per_image = n_patches_per_image,
            reduction_axes= 'C',
            target_axes   = 'YX',
            patch_axes    = None,
            save_file     = folder_out+'/training-data.npz',
            transforms    = None,
            patch_filter  = None,
            shuffle       = True,
            verbose       = True,
        )

I've played around with reshaping images in "2" to 3D (from (Y,X) to (1,Y,X)), changing the axes in RawData.from_folder(), changing the target_axes to 'CYX', the patch_size to (100,100), and so on. But always, errors arise (_ValueError: axes (CYX) must be of length 2._ / _raise e_

So the above settings are the only ones that seem to work for the data generation. But then, if I do the training, it flags the following: _ValueError: training images must be evenly divisible by 4 along axis C (which has incompatible size 3)_

Anyone know what settings are wrong? Or does this method really only work for data that is at least size 4 along C?

Thanks a lot in advance!
Lisa

**Comment by maweigert** [comment]
Hi Lisa, 

Hmm, I cannot reproduce this. 
Could you write in more detail what is the shape of your input files, what you exactly did, and what error was thrown?

**Comment by uschmidt83** [comment]
> first of all many thanks for creating such an easy-to-use framework for deep learning, I really appreciate it!

Thanks!

> I have tried to train the network with 3D images as source/low, and 2D images as target/GT, but have trouble making it work.

What exactly do you want to do? We have an [example for 3D -> 2D surface projection](http://csbdeep.bioimagecomputing.com/examples/projection/), but it seems that you want to do something with a timelapse?

I have the suspicion that you want to do something that is not supported by our software.

> I thought this is what create_patches_reduced_target() is for but I might have misunderstood it.

> So the above settings are the only ones that seem to work for the data generation.

It is and you seem to got it to work.

> But then, if I do the training, it flags the following: _ValueError: training images must be evenly divisible by 4 along axis T (which has incompatible size 3)_
> 
> Anyone know what settings are wrong? Or does this method really only work for data that is at least size 4 along T?

Are you using a `CARE` or `ProjectionCARE` model? As Martin already said, it would be easier to help if you provided more details what do you did.

Best,
Uwe

**Comment by lizard24** [comment]
Hi both,

many apologies for the late reply!

Ok, I went back into the examples and tried the projection example (which I hadn't tried before). When I change the axes in RawData.from_folder to "TYX" (instead of "ZXY") and the reduction_axes in create_patches_reduced_target to "T" (not "Z"), it does give me a value error when I want to start the training ("ValueError: projection axis Z not part of model axes TYXC").

Basically, my input data is a time series of an event (the GT data is one binary image), and instead of picking one image for the segmentation, I'd like to show the network the three images I took, to have as much information as possible for the decision. Could I use the projection example for this by keeping the default settings?

Thank you for your help!
Lisa

**Comment by uschmidt83** [comment]
Hi Lisa,

> Basically, my input data is a time series of an event (the GT data is one binary image), and instead of picking one image for the segmentation, I'd like to show the network the three images I took, to have as much information as possible for the decision.

that seems like a good idea. However, CARE is really meant for image restoration (output is an _intensity_ per pixel), not image segmentation (output is a foreground _probability_ per pixel) as you want to do.

> Could I use the projection example for this by keeping the default settings?

The `ProjectionCARE` model is quite specific and very likely not good for your application.

What you can do is to concatenate the 3 images to make a multi-channel image (time -> channel) and use that as the input to a `CARE` network. But as I said, you want to do image segmentation which requires different options (loss function, etc.) as in our examples.

If you send me some data (i.e. the 3 input images with the corresponding binary output image), I'll try to make an example to get you started.

Best,
Uwe

**Comment by lizard24** [comment]
Hi Uwe,

thanks for your answer!

Concatenating is a good idea, I will try that! Also, I will try to train with GT data, that has the same image along "T".

I understand that CARE is mainly for image restoration, but first tries to use it for other projects worked quite well.

I will play around with the data further and come back to you! Not ready to give up yet, especially with that much time in hand atm ;-)
Many thanks again,
Lisa
